File: python/examples/composable_ciphers.py
"""
Example demonstrating the composable cipher system.

This example shows how monoalphabetic ciphers can produce custom alphabets
that are then used by polygraphic ciphers, creating a powerful composable system.
"""

import sys
import os

# Add the parent directory to the path so we can import cryptology
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

# Import monoalphabetic ciphers
from cryptology.classical.substitution.monoalphabetic.caesar import produce_alphabet as caesar_produce
from cryptology.classical.substitution.monoalphabetic.keyword import produce_alphabet as keyword_produce
from cryptology.classical.substitution.monoalphabetic.affine import produce_alphabet as affine_produce
from cryptology.classical.substitution.monoalphabetic.atbash import produce_alphabet as atbash_produce

# Import polygraphic ciphers
from cryptology.classical.substitution.polygraphic.playfair import encrypt as playfair_encrypt, decrypt as playfair_decrypt
from cryptology.classical.substitution.polygraphic.two_square import encrypt as two_square_encrypt, decrypt as two_square_decrypt
from cryptology.classical.substitution.polygraphic.four_square import encrypt as four_square_encrypt, decrypt as four_square_decrypt
from cryptology.classical.substitution.polygraphic.hill import encrypt as hill_encrypt, decrypt as hill_decrypt


def demonstrate_caesar_playfair():
    """Demonstrate Caesar-produced alphabet with Playfair cipher."""
    print("=== Caesar + Playfair ===")
    
    plaintext = "HELLO WORLD"
    key = "MONARCHY"
    
    # Produce Caesar-shifted alphabet
    caesared_alphabet = caesar_produce(shift=5)
    print(f"Caesar-produced alphabet (shift=5): {caesared_alphabet}")
    print(f"Plaintext: {plaintext}")
    print(f"Playfair key: {key}")
    
    # Use Caesar-produced alphabet with Playfair
    encrypted = playfair_encrypt(plaintext, key, caesared_alphabet)
    decrypted = playfair_decrypt(encrypted, key, caesared_alphabet)
    
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")
    print()


def demonstrate_keyword_two_square():
    """Demonstrate Keyword-produced alphabet with Two Square cipher."""
    print("=== Keyword + Two Square ===")
    
    plaintext = "ATTACK AT DAWN"
    keyword1 = "SECRET"
    keyword2 = "PLAYFAIR"
    
    # Produce keyword-based alphabets
    keyword_alphabet1 = keyword_produce(keyword1)
    keyword_alphabet2 = keyword_produce(keyword2)
    
    print(f"Keyword-produced alphabet 1: {keyword_alphabet1}")
    print(f"Keyword-produced alphabet 2: {keyword_alphabet2}")
    print(f"Plaintext: {plaintext}")
    
    # Use keyword-produced alphabets with Two Square
    encrypted = two_square_encrypt(plaintext, keyword1, keyword2, keyword_alphabet1, keyword_alphabet2)
    decrypted = two_square_decrypt(encrypted, keyword1, keyword2, keyword_alphabet1, keyword_alphabet2)
    
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")
    print()


def demonstrate_affine_four_square():
    """Demonstrate Affine-produced alphabet with Four Square cipher."""
    print("=== Affine + Four Square ===")
    
    plaintext = "THE QUICK BROWN FOX"
    keys = ["MONARCHY", "PLAYFAIR", "CIPHER", "SECRET"]
    
    # Produce affine-transformed alphabets
    affine_alphabets = []
    for i, key in enumerate(keys):
        a, b = 5 + i, 8 + i  # Different affine parameters
        affine_alphabet = affine_produce(a, b)
        affine_alphabets.append(affine_alphabet)
        print(f"Affine-produced alphabet {i+1} (a={a}, b={b}): {affine_alphabet}")
    
    print(f"Plaintext: {plaintext}")
    
    # Use affine-produced alphabets with Four Square
    encrypted = four_square_encrypt(plaintext, *keys, *affine_alphabets)
    decrypted = four_square_decrypt(encrypted, *keys, *affine_alphabets)
    
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")
    print()


def demonstrate_atbash_hill():
    """Demonstrate Atbash-produced alphabet with Hill cipher."""
    print("=== Atbash + Hill ===")
    
    plaintext = "HELLO WORLD"
    hill_matrix = [[3, 3], [2, 5]]
    
    # Produce Atbash-reversed alphabet
    atbash_alphabet = atbash_produce()
    print(f"Atbash-produced alphabet: {atbash_alphabet}")
    print(f"Plaintext: {plaintext}")
    print(f"Hill matrix: {hill_matrix}")
    
    # Use Atbash-produced alphabet with Hill
    encrypted = hill_encrypt(plaintext, hill_matrix)
    decrypted = hill_decrypt(encrypted, hill_matrix)
    
    print(f"Encrypted: {encrypted}")
    print(f"Decrypted: {decrypted}")
    print()


def demonstrate_multi_layer_encryption():
    """Demonstrate multi-layer encryption using different combinations."""
    print("=== Multi-Layer Encryption ===")
    
    plaintext = "SECRET MESSAGE"
    
    # Layer 1: Caesar + Playfair
    caesared_alphabet = caesar_produce(shift=7)
    layer1_encrypted = playfair_encrypt(plaintext, "KEY1", caesared_alphabet)
    
    # Layer 2: Keyword + Two Square
    keyword_alphabet1 = keyword_produce("SECRET")
    keyword_alphabet2 = keyword_produce("KEY2")
    layer2_encrypted = two_square_encrypt(layer1_encrypted, "KEY1", "KEY2", keyword_alphabet1, keyword_alphabet2)
    
    # Layer 3: Affine + Hill
    affine_alphabet = affine_produce(5, 8)
    hill_matrix = [[3, 3], [2, 5]]
    layer3_encrypted = hill_encrypt(layer2_encrypted, hill_matrix)
    
    print(f"Original: {plaintext}")
    print(f"Layer 1 (Caesar+Playfair): {layer1_encrypted}")
    print(f"Layer 2 (Keyword+Two Square): {layer2_encrypted}")
    print(f"Layer 3 (Affine+Hill): {layer3_encrypted}")
    print()


def demonstrate_custom_language_combinations():
    """Demonstrate combinations with custom languages."""
    print("=== Custom Language Combinations ===")
    
    # Turkish alphabet
    turkish_alphabet = "abcçdefgğhıijklmnoöprsştuüvyz"
    plaintext = "merhaba dünya"
    key = "gizli"
    
    # Produce Caesar-shifted Turkish alphabet
    caesared_turkish = caesar_produce(shift=3, alphabet=turkish_alphabet)
    print(f"Turkish alphabet: {turkish_alphabet}")
    print(f"Caesar-shifted Turkish: {caesared_turkish}")
    print(f"Plaintext: {plaintext}")
    
    # Use with Playfair
    try:
        encrypted = playfair_encrypt(plaintext, key, caesared_turkish)
        decrypted = playfair_decrypt(encrypted, key, caesared_turkish)
        print(f"Encrypted: {encrypted}")
        print(f"Decrypted: {decrypted}")
    except Exception as e:
        print(f"Error: {e}")
    
    print()


def demonstrate_alphabet_analysis():
    """Demonstrate analysis of produced alphabets."""
    print("=== Alphabet Analysis ===")
    
    base_alphabet = "abcdefghijklmnopqrstuvwxyz"
    
    # Different Caesar shifts
    for shift in [1, 3, 5, 13]:
        caesared = caesar_produce(shift, base_alphabet)
        print(f"Caesar shift {shift:2d}: {caesared}")
    
    print()
    
    # Different keywords
    keywords = ["SECRET", "PLAYFAIR", "CIPHER"]
    for keyword in keywords:
        keyword_alphabet = keyword_produce(keyword, base_alphabet)
        print(f"Keyword '{keyword}': {keyword_alphabet}")
    
    print()
    
    # Different affine parameters
    affine_params = [(3, 1), (5, 8), (7, 13)]
    for a, b in affine_params:
        try:
            affine_alphabet = affine_produce(a, b, base_alphabet)
            print(f"Affine (a={a}, b={b}): {affine_alphabet}")
        except ValueError as e:
            print(f"Affine (a={a}, b={b}): Error - {e}")
    
    print()


if __name__ == "__main__":
    print("Composable Cipher System Demo")
    print("=" * 50)
    print()
    
    demonstrate_caesar_playfair()
    demonstrate_keyword_two_square()
    demonstrate_affine_four_square()
    demonstrate_atbash_hill()
    demonstrate_multi_layer_encryption()
    demonstrate_custom_language_combinations()
    demonstrate_alphabet_analysis()
    
    print("Demo completed!")
    print()
    print("Key Benefits of Composable System:")
    print("1. Enhanced Security: Multiple layers of encryption")
    print("2. Flexibility: Mix and match different cipher types")
    print("3. Customization: Use any alphabet with any cipher")
    print("4. Educational: Understand how different ciphers interact")
    print("5. Practical: Real-world applications often use multiple ciphers")

File: c/examples/composable_ciphers.c
/**
 * @file composable_ciphers.c
 * @brief Example demonstrating the composable cipher system
 * 
 * This example shows how monoalphabetic ciphers can produce custom alphabets
 * that are then used with polygraphic ciphers for enhanced security.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

// Include monoalphabetic cipher headers
#include "cryptology/classical/substitution/monoalphabetic/caesar.h"
#include "cryptology/classical/substitution/monoalphabetic/keyword.h"
#include "cryptology/classical/substitution/monoalphabetic/affine.h"
#include "cryptology/classical/substitution/monoalphabetic/atbash.h"

// Include polygraphic cipher headers
#include "cryptology/classical/substitution/polygraphic/playfair.h"

#define MAX_BUFFER_SIZE 1024
#define DEFAULT_ALPHABET "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
#define TURKISH_ALPHABET "ABCÇDEFGĞHIJKLMNOÖPRSŞTUÜVYZ"

void demonstrate_caesar_playfair() {
    printf("=== Caesar + Playfair Composable System ===\n");
    
    const char *plaintext = "HELLO WORLD";
    const char *playfair_key = "MONARCHY";
    int caesar_shift = 5;
    
    char caesar_alphabet[MAX_BUFFER_SIZE];
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    
    printf("Plaintext: %s\n", plaintext);
    printf("Caesar shift: %d\n", caesar_shift);
    printf("Playfair key: %s\n", playfair_key);
    printf("\n");
    
    // Step 1: Produce Caesar-shifted alphabet
    if (caesar_produce_alphabet(caesar_shift, DEFAULT_ALPHABET, caesar_alphabet, sizeof(caesar_alphabet)) != 0) {
        printf("Error: Failed to produce Caesar alphabet\n");
        return;
    }
    printf("Caesar-produced alphabet: %s\n", caesar_alphabet);
    
    // Step 2: Use Caesar alphabet with Playfair
    if (playfair_encrypt_with_alphabet(plaintext, playfair_key, caesar_alphabet, encrypted, sizeof(encrypted)) != 0) {
        printf("Error: Failed to encrypt with Playfair\n");
        return;
    }
    printf("Encrypted: %s\n", encrypted);
    
    // Step 3: Decrypt with same alphabet
    if (playfair_decrypt_with_alphabet(encrypted, playfair_key, caesar_alphabet, decrypted, sizeof(decrypted)) != 0) {
        printf("Error: Failed to decrypt with Playfair\n");
        return;
    }
    printf("Decrypted: %s\n", decrypted);
    printf("\n");
}

void demonstrate_keyword_playfair() {
    printf("=== Keyword + Playfair Composable System ===\n");
    
    const char *plaintext = "SECRET MESSAGE";
    const char *keyword = "SECRET";
    const char *playfair_key = "MONARCHY";
    
    char keyword_alphabet[MAX_BUFFER_SIZE];
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    
    printf("Plaintext: %s\n", plaintext);
    printf("Keyword: %s\n", keyword);
    printf("Playfair key: %s\n", playfair_key);
    printf("\n");
    
    // Step 1: Produce keyword-based alphabet
    if (keyword_produce_alphabet(keyword, DEFAULT_ALPHABET, keyword_alphabet, sizeof(keyword_alphabet)) != 0) {
        printf("Error: Failed to produce keyword alphabet\n");
        return;
    }
    printf("Keyword-produced alphabet: %s\n", keyword_alphabet);
    
    // Step 2: Use keyword alphabet with Playfair
    if (playfair_encrypt_with_alphabet(plaintext, playfair_key, keyword_alphabet, encrypted, sizeof(encrypted)) != 0) {
        printf("Error: Failed to encrypt with Playfair\n");
        return;
    }
    printf("Encrypted: %s\n", encrypted);
    
    // Step 3: Decrypt with same alphabet
    if (playfair_decrypt_with_alphabet(encrypted, playfair_key, keyword_alphabet, decrypted, sizeof(decrypted)) != 0) {
        printf("Error: Failed to decrypt with Playfair\n");
        return;
    }
    printf("Decrypted: %s\n", decrypted);
    printf("\n");
}

void demonstrate_affine_playfair() {
    printf("=== Affine + Playfair Composable System ===\n");
    
    const char *plaintext = "AFFINE CIPHER";
    int a = 3, b = 5;  // Affine parameters
    const char *playfair_key = "MONARCHY";
    
    char affine_alphabet[MAX_BUFFER_SIZE];
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    
    printf("Plaintext: %s\n", plaintext);
    printf("Affine parameters: a=%d, b=%d\n", a, b);
    printf("Playfair key: %s\n", playfair_key);
    printf("\n");
    
    // Step 1: Produce affine-transformed alphabet
    if (affine_produce_alphabet(a, b, DEFAULT_ALPHABET, affine_alphabet, sizeof(affine_alphabet)) != 0) {
        printf("Error: Failed to produce affine alphabet\n");
        return;
    }
    printf("Affine-produced alphabet: %s\n", affine_alphabet);
    
    // Step 2: Use affine alphabet with Playfair
    if (playfair_encrypt_with_alphabet(plaintext, playfair_key, affine_alphabet, encrypted, sizeof(encrypted)) != 0) {
        printf("Error: Failed to encrypt with Playfair\n");
        return;
    }
    printf("Encrypted: %s\n", encrypted);
    
    // Step 3: Decrypt with same alphabet
    if (playfair_decrypt_with_alphabet(encrypted, playfair_key, affine_alphabet, decrypted, sizeof(decrypted)) != 0) {
        printf("Error: Failed to decrypt with Playfair\n");
        return;
    }
    printf("Decrypted: %s\n", decrypted);
    printf("\n");
}

void demonstrate_atbash_playfair() {
    printf("=== Atbash + Playfair Composable System ===\n");
    
    const char *plaintext = "ATBASH CIPHER";
    const char *playfair_key = "MONARCHY";
    
    char atbash_alphabet[MAX_BUFFER_SIZE];
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    
    printf("Plaintext: %s\n", plaintext);
    printf("Playfair key: %s\n", playfair_key);
    printf("\n");
    
    // Step 1: Produce Atbash-reversed alphabet
    if (atbash_produce_alphabet(DEFAULT_ALPHABET, atbash_alphabet, sizeof(atbash_alphabet)) != 0) {
        printf("Error: Failed to produce Atbash alphabet\n");
        return;
    }
    printf("Atbash-produced alphabet: %s\n", atbash_alphabet);
    
    // Step 2: Use Atbash alphabet with Playfair
    if (playfair_encrypt_with_alphabet(plaintext, playfair_key, atbash_alphabet, encrypted, sizeof(encrypted)) != 0) {
        printf("Error: Failed to encrypt with Playfair\n");
        return;
    }
    printf("Encrypted: %s\n", encrypted);
    
    // Step 3: Decrypt with same alphabet
    if (playfair_decrypt_with_alphabet(encrypted, playfair_key, atbash_alphabet, decrypted, sizeof(decrypted)) != 0) {
        printf("Error: Failed to decrypt with Playfair\n");
        return;
    }
    printf("Decrypted: %s\n", decrypted);
    printf("\n");
}

void demonstrate_turkish_playfair() {
    printf("=== Turkish Alphabet + Playfair ===\n");
    
    const char *plaintext = "MERHABA DÜNYA";
    const char *playfair_key = "GİZLİ";
    
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    
    printf("Plaintext: %s\n", plaintext);
    printf("Playfair key: %s\n", playfair_key);
    printf("Turkish alphabet: %s\n", TURKISH_ALPHABET);
    printf("\n");
    
    // Use Turkish alphabet directly with Playfair
    if (playfair_encrypt_with_alphabet(plaintext, playfair_key, TURKISH_ALPHABET, encrypted, sizeof(encrypted)) != 0) {
        printf("Error: Failed to encrypt with Playfair using Turkish alphabet\n");
        return;
    }
    printf("Encrypted: %s\n", encrypted);
    
    // Decrypt with same alphabet
    if (playfair_decrypt_with_alphabet(encrypted, playfair_key, TURKISH_ALPHABET, decrypted, sizeof(decrypted)) != 0) {
        printf("Error: Failed to decrypt with Playfair using Turkish alphabet\n");
        return;
    }
    printf("Decrypted: %s\n", decrypted);
    printf("\n");
}

void demonstrate_multi_layer_encryption() {
    printf("=== Multi-Layer Encryption ===\n");
    
    const char *plaintext = "MULTI LAYER";
    const char *keyword = "SECRET";
    int caesar_shift = 3;
    const char *playfair_key = "MONARCHY";
    
    char keyword_alphabet[MAX_BUFFER_SIZE];
    char caesar_alphabet[MAX_BUFFER_SIZE];
    char final_alphabet[MAX_BUFFER_SIZE];
    char encrypted[MAX_BUFFER_SIZE];
    char decrypted[MAX_BUFFER_SIZE];
    
    printf("Plaintext: %s\n", plaintext);
    printf("Layer 1 - Keyword: %s\n", keyword);
    printf("Layer 2 - Caesar shift: %d\n", caesar_shift);
    printf("Layer 3 - Playfair key: %s\n", playfair_key);
    printf("\n");
    
    // Layer 1: Keyword alphabet
    if (keyword_produce_alphabet(keyword, DEFAULT_ALPHABET, keyword_alphabet, sizeof(keyword_alphabet)) != 0) {
        printf("Error: Failed to produce keyword alphabet\n");
        return;
    }
    printf("Keyword alphabet: %s\n", keyword_alphabet);
    
    // Layer 2: Caesar shift the keyword alphabet
    if (caesar_produce_alphabet(caesar_shift, keyword_alphabet, caesar_alphabet, sizeof(caesar_alphabet)) != 0) {
        printf("Error: Failed to produce Caesar alphabet\n");
        return;
    }
    printf("Caesar-shifted alphabet: %s\n", caesar_alphabet);
    
    // Layer 3: Use with Playfair
    if (playfair_encrypt_with_alphabet(plaintext, playfair_key, caesar_alphabet, encrypted, sizeof(encrypted)) != 0) {
        printf("Error: Failed to encrypt with Playfair\n");
        return;
    }
    printf("Final encrypted: %s\n", encrypted);
    
    // Decrypt (reverse the process)
    if (playfair_decrypt_with_alphabet(encrypted, playfair_key, caesar_alphabet, decrypted, sizeof(decrypted)) != 0) {
        printf("Error: Failed to decrypt with Playfair\n");
        return;
    }
    printf("Decrypted: %s\n", decrypted);
    printf("\n");
}

int main() {
    printf("Composable Cipher System Demo\n");
    printf("=============================\n\n");
    
    demonstrate_caesar_playfair();
    demonstrate_keyword_playfair();
    demonstrate_affine_playfair();
    demonstrate_atbash_playfair();
    demonstrate_turkish_playfair();
    demonstrate_multi_layer_encryption();
    
    printf("Summary:\n");
    printf("1. Monoalphabetic ciphers can produce custom alphabets\n");
    printf("2. Polygraphic ciphers can use these custom alphabets\n");
    printf("3. Multiple layers of encryption for enhanced security\n");
    printf("4. Support for different languages (English, Turkish)\n");
    printf("5. Composable system provides flexible encryption options\n");
    
    return 0;
}
